core/src/main/java/org/bitcoinj/crypto/HDPath.java

```diff
@@ -444,13 +456,42 @@ public List<HDPath> ancestors() {
      * @return unmodifiable list of ancestors
      */
     public List<HDPath> ancestors(boolean includeSelf) {
-        int endExclusive =  childNumbers.size() + (includeSelf ? 1 : 0);
-        return IntStream.range(1, endExclusive)
-                .mapToObj(i -> childNumbers.subList(0, i))
-                .map(l -> HDPath.of(((HDFullPath) this).hasPrivateKey, l))
+        int endExclusive =  childNumbers.size() + (includeSelf ? 0 : -1);
+        return IntStream.range(0, endExclusive)
+                .mapToObj(this::ancestorByIndex)
                 .collect(StreamUtils.toUnmodifiableList());
     }
 
+    public HDPath ancestorByIndex(int index) {
+        checkArgument(index >= 0 && index < childNumbers.size(), () -> String.format("Index %s out of bounds (0, %s)", index, childNumbers.size() - 1));
+        List<ChildNumber> subList = subListInternal(index + 1);
+        return cloneWithPath(subList);
+    }
+
+    // Depth of 0, returns empty path, Depth of d returns ancestorByIndex(d-1)
+    public HDPath ancestorByDepth(int depth) {
+        checkArgument(depth >= 0 && depth <= childNumbers.size(), () -> String.format("Depth %s out of bounds (0, %s)", depth, childNumbers.size()));
+        return depth == 0
+                ? cloneEmpty()
+                : this.ancestorByIndex(depth - 1);
+    }
+
+    private List<ChildNumber> subListInternal(int to) {
+        return childNumbers.subList(0, to);
+    }
```

Comment:
This refactoring (including the usages) could perhaps be in its own commit, to clear this one up a bit.

Reply:
Fixed.